README.md
Taken from ubuntu 2004 arm repos. The yoga c630 arm6 laptop does not survive
the cpu wiggle max_Scaling_frequency reapply, so I didabled it. So far so good.
Without this patch my cpu speeds would fall and stay at the lowest possible
freqs.
In order to build natively I had to uninstall autoconf-archive to get around
"too many loops" error and I had to replace the pthread check in
configure.ac with CXXFLAGS, I left the original as configure.ac-og if your
build enviroment is better then mine and want to change it.


﻿Build dependencies
------------------
PowerTOP uses C++, and expects g++ and libstdc++ to be functional
along with a glibc/pthreadsi, autoconf, automake, and libtool in the
development environment.

In addition to that, PowerTOP requires the following components:

pciutils-devel (is only required if you have PCI)
ncurses-devel  (required)
libnl-devel    (required)
kernel version => 2.6.38

Example packages to install in Ubuntu

sudo apt-get install libpci-dev
sudo apt-get install libnl-3-dev libnl-genl-3-dev
sudo apt-get install gettext libgettextpo-dev
sudo apt-get install autopoint
sudo apt-get install gettext
sudo apt-get install libncurses5-dev libncursesw5-dev
sudo apt-get install libtool-bin
sudo apt-get install dh-autoreconf

Building & Installing PowerTOP
------------------------------

To build and install PowerTOP type the following commands,
       ./autogen.sh
       ./configure
       ./make
       ./make install

The autogen script is required to run only once to generate the configure for
the build. Rest of the steps are required to run everytime there is a change
made to the PowerTop.

Note: For Android (running Intel Architecture ) there is a Android.mk
that was provided by community members, and at this time is supported
mostly by community members.

Build PowerTop for Android LollyPop
------------------------------------

1. Obtain pciutils and ncurses libraries.
   Skip this step if they are already in the tree.

   pciutils: Clone
       https://github.com/trevd/android_external_pciutils
   into external/pciutils

   ncurses: Clone
       https://github.com/cvpcs/android_external_libncurses
   into external/ncurses

2. Run
       ./src/csstoh.sh src/powertop.css src/css.h
   to generate the css.h header file.

3. Apply patches inside patches/Android

4. PowerTop can be built with top level make,
   or by doing "mm".

   The resulting binary is under $ANDROID_PRODUCT_OUT/system/bin/powertop


Kernel Parameters:
------------------

PowerTOP needs some kernel config options enabled in order function properly.
As of linux-3.3.0 these are (list probably incomplete):

CONFIG_NO_HZ
CONFIG_HIGH_RES_TIMERS
CONFIG_HPET_TIMER
CONFIG_CPU_FREQ_GOV_ONDEMAND
CONFIG_USB_SUSPEND
CONFIG_SND_AC97_POWER_SAVE
CONFIG_TIMER_STATS
CONFIG_PERF_EVENTS
CONFIG_PERF_COUNTERS
CONFIG_TRACEPOINTS
CONFIG_TRACING
CONFIG_EVENT_POWER_TRACING_DEPRECATED
CONFIG_X86_MSR
ACPI_PROCFS_POWER
CONFIG_DEBUG_FS

Use these configs from linux-3.13.rc1
CONFIG_POWERCAP
CONFIG_INTEL_RAPL

The patches in the patches/ sub-directory are required for PowerTOP to function
fully.



Outputting a report
-------------------
When invoking PowerTOP without arguments, PowerTOP starts in interactive mode.
However, for reporting bugs etc there is a special reporting modes:

For an HTML formatted report simply execute PowerTOP with the following,

powertop --html

which will create a "powertop.html" file which is static and can be sent to
others to help diagnose power issues. Additionally has the feature of showing
the tunables executions within the report as a reference.

Note for the developers: If you make changes on the report format please
make sure that the output can be cleanly validated by the W3C Markup
Validation Service and W3C CSS Validation Service:
 * http://validator.w3.org/#validate_by_upload
 * http://jigsaw.w3.org/css-validator/#validate_by_upload

For a CSV formatted report, simply execute PowerTOP with the following,

Powertop --csv

which will create a “powertop.csv” file which is static and can be used in
reporting, diagnoses, and analytical data analysis.

Also you can set the number of iterations, and duration of test, in which case
all reports will time stamp for you.

Calibrating & Power Numbers
---------------------------
PowerTOP will, when running on battery, track your power consumption as well
as your activity on the system. Once there are sufficient such measurements,
PowerTOP can start to report power estimates for various activities.
You can help get this estimation more accurate by running a calibration
cycle:

powertop --calibrate

at least once; this will cycle through various display brightness levels
(including "off") as well as USB device activities and some other workloads.


Code from other open source projects
------------------------------------
PowerTOP contains some code from other open source projects; we'd like to
thank the authors of those projects for their work.
Specifically PowerTOP contains code from

Parse Event Library - Copyright 2009, 2010 Red Hat Inc  Steven Rosted <srostedt@redhat.com>
nl80211 userspace tool - Copyright 2007, 2008	Johannes Berg <johannes@sipsolutions.net>


Extech Power Analyzer / Datalogger support
------------------------------------------
I use, and our analysis teams use, the Extech Power Analyzer/Datalogger
(model number 380803) quite a lot, and PowerTOP supports using this
device over the serial cable. Just pass the device node on the command line
like this

powertop --extech=/dev/ttyUSB0

(where ttyUSB0 is the devicenode of the serial-to-usb adapter on my system)



debian/changelog
powertop (2.11-1build4) focal; urgency=medium

  * patches for yoga c630

 -- Ben Maddocks <bm16ton@gmail.com>  Tue, 17 Nov 2020 13:29:06 +0100

powertop (2.11-1build1) focal; urgency=medium

  * No-change rebuild for libgcc-s1 package name change.

 -- Matthias Klose <doko@ubuntu.com>  Sat, 21 Mar 2020 13:29:06 +0100

powertop (2.11-1) unstable; urgency=medium

  [ Kan-Ru Chen ]
  * New upstream version 2.11
  * debian/README.Debian: Added to explain how parameters collection works
    (Closes: #823701)
  * debian/patches: Add patch to use correct hciconfig binary path
    (Closes: #772571)

 -- Kan-Ru Chen (陳侃如) <koster@debian.org>  Fri, 03 Jan 2020 23:37:27 +0900

powertop (2.10-1) unstable; urgency=medium

  [ Julian Wollrath ]
  * New upstream release (Closes: #732458, #793275, #857941):
    - Support POWER (Closes: #463034, #807610).
    - Update debian/copyright.
    - USB devices, that cannot be autosuspended are not shown anymore
      (Closes: #664455).
  * Bump standards version to 3.9.8 (no changes needed).
  * Drop manually generated debug package.
  * Update watchfile to look at GitHub.
  * Remove myself from uploaders.

  [ Ross Vandegrift ]
  * d/watch: fix filemangle
  * Avoid linking with ncurses (Thanks to Sven Joachim for this patch)
    (Closes: #898533)
  * Switch to dh compat 11
  * d/control:
    + Set Rules-Requires-Root to no
    + Bump Standards-Version to 4.3.0 (no changes required)
  * d/rules:
    + Enable build-time hardening
    + Drop unnecessary dh_installdocs override
  * d/gbp.conf: enable pristine-tar by default

  [ Kan-Ru Chen ]
  * New upstream version 2.10
  * d/control: point Vcs-* field to Salsa
  * d/control: Add myself and Ross Vandegrift as uploaders, remove
    Jose Luis Rivas and Patrick Winnertz (Closes: #931713)
  * d/watch: Use official tarball from upstream website
  * d/patches: drop 0001-Add-version-files-for-v2.9.patch

 -- Kan-Ru Chen (陳侃如) <koster@debian.org>  Fri, 02 Aug 2019 22:15:56 +0900

powertop (2.8-1) unstable; urgency=medium

  * new upstream release (Closes: #801873)

 -- Jose Luis Rivas <ghostbar@debian.org>  Fri, 06 Nov 2015 12:38:09 -0500

powertop (2.7-1) unstable; urgency=medium

  * New upstream version to unstable. (Closes: #795297)

 -- Jose Luis Rivas <ghostbar@debian.org>  Wed, 09 Sep 2015 08:41:21 -0500

powertop (2.7-1~exp1) experimental; urgency=medium

  * Update debian/watch (modified the one provided by bartm, thanks).
  * New upstream release:
    + RAPL: powercap sysfs used for power measurements.
    + ARM support for devfreq stats.
    + Added support to new Intel platforms.
    + Several bug fixes.
  * Update debian/copyright.
  * Bump standards version to 3.9.6, no changes needed.

 -- Julian Wollrath <jwollrath@web.de>  Mon, 24 Nov 2014 23:23:37 +0100

powertop (2.6.1-1) unstable; urgency=medium

  * debian/watch: adjust to new URL.
  * Change architectures to linux-any, since ia64 is not a release
    architecture anymore.
  * New upstream release.
    + Drop now obsolete patch.
  * Update year in debian/copyright.
  * Drop automake from B-D (pulled in by dh-autoreconf anyways).

 -- Julian Wollrath <jwollrath@web.de>  Wed, 23 Jul 2014 16:28:08 +0200

powertop (2.5-1) unstable; urgency=low

  * Only build for supported architectures.
  * New upstream release (Closes: #695890, #726539).
    + Use dh-autoreconf to make it build with automake 1.14.
    + Add patch to not use csstoh.sh.
  * Bump standards version to 3.9.5, no changes needed.

 -- Julian Wollrath <jwollrath@web.de>  Sat, 23 Nov 2013 14:02:00 +0100

powertop (2.4-2) unstable; urgency=low

  * debian/copyright: Add missing copyright notices.
  * Add myself as co-maintainer.
  * Add postrm script to remove /var/cache/powertop when purging
    (Closes: #680510).
  * Change architecture to linux-any, since it is a linux specific package.
  * Change package description to something more descriptive.
  * Add debug package (Closes: #679579).
  * Remove unneeded copy of the manpage, use the upstream one instead.
  * Disable silent rules during build.

 -- Julian Wollrath <jwollrath@web.de>  Mon, 02 Sep 2013 17:36:42 -0400

powertop (2.4-1) unstable; urgency=low

  * New upstream release (Closes: #685607, #695893, #659517, #686699, #705993)
  * debian/control:
    + Jose Luis Rivas's email to `ghostbar@debian.org`.
    + Remove build-dependency on `cdbs`. No need anymore.
    + Add dependency on `automake` so `config.{guess,sub}` are auto-updated on
    build.
    + Standards-Version to 3.9.4.
    + Swap on fields maintainer and uploaders.
    + Add Vcs-* fields.
  * debian/watch: upstream changed to `tar.gz` from `tar.bz2`, so reports on
    newer versions were failing to be accurate. (Closes: #695891)
  * Acknowledging NMU's.
  * debian/rules: Simple dh format.
  * debian/patches: changes applied upstream. No need anymore.
  * debian/copyright: update to copyright-format/1.0

 -- Jose Luis Rivas <ghostbar@debian.org>  Fri, 30 Aug 2013 22:47:26 -0430

powertop (2.0-0.3) unstable; urgency=high

  * Non-maintainer upload.
  * Add patch long-process-names-nullterm.patch to prevent a crash
    on paticularly long process names (Closes: #682388)

 -- Jonathan Wiltshire <jmw@debian.org>  Sun, 21 Apr 2013 00:38:10 +0100

powertop (2.0-0.2) unstable; urgency=low

  * Non-maintainer upload.
  * Drop debian/patches/fix_libnl_version.patch. This patch actually doesn't
    make any sense and only causes a FTBFS on ia64 since it modifies
    configure.ac without properly updating the build system afterwards.
    (Closes: #679629)

 -- Michael Biebl <biebl@debian.org>  Sat, 30 Jun 2012 16:13:46 +0200

powertop (2.0-0.1) unstable; urgency=low

  * Non-maintainer upload.
  * New upstream version (Closes: #665695, #672555, #652466, #638300)
  * Build against libnl3 (Closes: #656421)
  * Add manpage, cherry-picked from upstream sources (Closes: #564847)
  * Update watch and copyright file (Closes: #650063, #634218)

src/tuning/bluetooth.cpp
/*
 * Copyright 2010, Intel Corporation
 *
 * This file is part of PowerTOP
 *
 * This program file is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program in a file named COPYING; if not, write to the
 * Free Software Foundation, Inc,
 * 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA
 * or just google for it.
 *
 * Authors:
 *	Arjan van de Ven <arjan@linux.intel.com>
 */

#include "tuning.h"
#include "tunable.h"
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <utility>
#include <iostream>
#include <fstream>
#include <unistd.h>
#include <sys/types.h>
#include <dirent.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <errno.h>

#include "../lib.h"
#include "bluetooth.h"

bt_tunable::bt_tunable(void) : tunable("", 1.0, _("Good"), _("Bad"), _("Unknown"))
{
	sprintf(desc, _("Bluetooth device interface status"));
	pt_strcpy(toggle_bad, "/bin/hciconfig hci0 up &> /dev/null &");
	pt_strcpy(toggle_good, "/bin/hciconfig hci0 down &> /dev/null");
}



/* structure definitions copied from include/net/bluetooth/hci.h from the 2.6.20 kernel */
#define HCIGETDEVINFO   _IOR('H', 211, int)
#define BTPROTO_HCI     1

#define __u16 uint16_t
#define __u8 uint8_t
#define __u32 uint32_t

typedef struct {
        __u8 b[6];
} __attribute__((packed)) bdaddr_t;

struct hci_dev_stats {
        __u32 err_rx;
        __u32 err_tx;
        __u32 cmd_tx;
        __u32 evt_rx;
        __u32 acl_tx;
        __u32 acl_rx;
        __u32 sco_tx;
        __u32 sco_rx;
        __u32 byte_rx;
        __u32 byte_tx;
};


struct hci_dev_info {
	__u16 dev_id;
	char  name[8];

	bdaddr_t bdaddr;

	__u32 flags;
	__u8  type;

	__u8  features[8];

	__u32 pkt_type;
	__u32 link_policy;
	__u32 link_mode;

	__u16 acl_mtu;
	__u16 acl_pkts;
	__u16 sco_mtu;
	__u16 sco_pkts;

	struct hci_dev_stats stat;
};

static int previous_bytes = -1;
static time_t last_check_time = 0;
static int last_check_result;

int bt_tunable::good_bad(void)
{
	struct hci_dev_info devinfo;
	FILE *file = 0;
	int fd;
	int thisbytes = 0;
	int ret;
	int result = TUNE_GOOD;

	fd = socket(AF_BLUETOOTH, SOCK_RAW, BTPROTO_HCI);
	if (fd < 0)
		return TUNE_GOOD;

	memset(&devinfo, 0, sizeof(devinfo));
	strcpy(devinfo.name, "hci0");
	ret = ioctl(fd, HCIGETDEVINFO, (void *) &devinfo);
	if (ret < 0)
		goto out;

	if ( (devinfo.flags & 1) == 0 &&
		access("/sys/module/hci_usb",F_OK)) /* interface down already */
		goto out;

	thisbytes += devinfo.stat.byte_rx;
	thisbytes += devinfo.stat.byte_tx;

	/* device is active... so we need to leave it on */
	if (thisbytes != previous_bytes)
		goto out;


	/* this check is expensive.. only do it once per minute */
	if (time(NULL) - last_check_time > 60) {
		last_check_result = TUNE_BAD;
		/* now, also check for active connections */
		file = popen("/usr/bin/hcitool con 2> /dev/null", "r");
		if (file) {
			char line[2048];
			/* first line is standard header */
			if (fgets(line, 2047, file) == NULL)
				goto out;
			memset(line, 0, 2048);
			if (fgets(line, 2047, file) == NULL) {
				result = last_check_result = TUNE_GOOD;
				goto out;
			}

			if (strlen(line) > 0) {
				result = last_check_result = TUNE_GOOD;
				goto out;
			}
		}
		last_check_time = time(NULL);
	};

	result = last_check_result;

out:
	previous_bytes = thisbytes;
	if (file)
		pclose(file);
	close(fd);
	return result;
}

void bt_tunable::toggle(void)
{
	int good;
	good = good_bad();

	if (good == TUNE_GOOD) {
		if(system("/bin/hciconfig hci0 up &> /dev/null &"))
			printf("System is not available\n");
		return;
	}
	if(system("/bin/hciconfig hci0 down &> /dev/null"))
		printf("System is not available\n");
}

const char *bt_tunable::toggle_script(void)
{
	int good;
	good = good_bad();

	if (good == TUNE_GOOD) {
		return toggle_bad;
	}
	return toggle_good;
}


void add_bt_tunable(void)
{
	struct hci_dev_info devinfo;
	class bt_tunable *bt;
	int fd;
	int ret;

	/* first check if /sys/modules/bluetooth exists, if not, don't probe bluetooth because
	   it would trigger an autoload */

//	if (access("/sys/module/bluetooth",F_OK))
//		return;


	/* check if hci0 exists */
	fd = socket(AF_BLUETOOTH, SOCK_RAW, BTPROTO_HCI);
	if (fd < 0)
		return;

	memset(&devinfo, 0, sizeof(devinfo));
	strcpy(devinfo.name, "hci0");
	ret = ioctl(fd, HCIGETDEVINFO, (void *) &devinfo);
	close(fd);
	if (ret < 0)
		return;


	bt = new class bt_tunable();
	all_tunables.push_back(bt);
}


src/main.cpp
/*
 * Copyright 2010, Intel Corporation
 *
 * This is part of PowerTOP
 *
 * This program file is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program in a file named COPYING; if not, write to the
 * Free Software Foundation, Inc,
 * 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA
 * or just google for it.
 *
 * getopt code is taken from "The GNU C Library" reference manual,
 * section 24.2 "Parsing program options using getopt"
 * http://www.gnu.org/s/libc/manual/html_node/Getopt-Long-Option-Example.html
 * Manual published under the terms of the Free Documentation License.
 *
 * Authors:
 *	Arjan van de Ven <arjan@linux.intel.com>
 */
#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <sys/stat.h>
#include <sys/statfs.h>
#include <getopt.h>
#include <unistd.h>
#include <locale.h>
#include <sys/resource.h>
#include <limits.h>
#include <pthread.h>

#include "cpu/cpu.h"
#include "process/process.h"
#include "perf/perf.h"
#include "perf/perf_bundle.h"
#include "lib.h"
#include "../config.h"


#include "devices/device.h"
#include "devices/devfreq.h"
#include "devices/usb.h"
#include "devices/ahci.h"
#include "measurement/measurement.h"
#include "parameters/parameters.h"
#include "calibrate/calibrate.h"


#include "tuning/tuning.h"
#include "wakeup/wakeup.h"

#include "display.h"
#include "devlist.h"
#include "report/report.h"

#define DEBUGFS_MAGIC          0x64626720

#define NR_OPEN_DEF 1024 * 1024

int debug_learning = 0;
unsigned time_out = 20;
int leave_powertop = 0;
void (*ui_notify_user) (const char *frmt, ...);

enum {
	OPT_AUTO_TUNE = CHAR_MAX + 1,
	OPT_EXTECH,
	OPT_DEBUG
};

static const struct option long_options[] =
{
	/* These options set a flag. */
	{"auto-tune",	no_argument,		NULL,		 OPT_AUTO_TUNE},
	{"calibrate",	no_argument,		NULL,		 'c'},
	{"csv",		optional_argument,	NULL,		 'C'},
	{"debug",	no_argument,		&debug_learning, OPT_DEBUG},
	{"extech",	optional_argument,	NULL,		 OPT_EXTECH},
	{"html",	optional_argument,	NULL,		 'r'},
	{"iteration",	optional_argument,	NULL,		 'i'},
	{"quiet",	no_argument,		NULL,		 'q'},
	{"sample",	optional_argument,	NULL,		 's'},
	{"time",	optional_argument,	NULL,		 't'},
	{"workload",	optional_argument,	NULL,		 'w'},
	{"version",	no_argument,		NULL,		 'V'},
	{"help",	no_argument,		NULL,		 'h'},
	{NULL,		0,			NULL,		 0}
};

static void print_version()
{
	printf(_("PowerTOP version " PACKAGE_VERSION "\n"));
}

static bool set_refresh_timeout()
{
	static char buf[4];
	mvprintw(1, 0, "%s (currently %u): ", _("Set refresh time out"), time_out);
	memset(buf, '\0', sizeof(buf));
	get_user_input(buf, sizeof(buf) - 1);
	show_tab(0);
	unsigned time = strtoul(buf, NULL, 0);
	if (!time) return 0;
	if (time > 32) time = 32;
	time_out = time;
	return 1;
}

static void print_usage()
{
	printf("%s\n\n", _("Usage: powertop [OPTIONS]"));
	printf("     --auto-tune\t %s\n", _("sets all tunable options to their GOOD setting"));
	printf(" -c, --calibrate\t %s\n", _("runs powertop in calibration mode"));
	printf(" -C, --csv%s\t %s\n", _("[=filename]"), _("generate a csv report"));
	printf("     --debug\t\t %s\n", _("run in \"debug\" mode"));
	printf("     --extech%s\t %s\n", _("[=devnode]"), _("uses an Extech Power Analyzer for measurements"));
	printf(" -r, --html%s\t %s\n", _("[=filename]"), _("generate a html report"));
	printf(" -i, --iteration%s\n", _("[=iterations] number of times to run each test"));
	printf(" -q, --quiet\t\t %s\n", _("suppress stderr output"));
	printf(" -s, --sample%s\t %s\n", _("[=seconds]"), _("interval for power consumption measurement"));
	printf(" -t, --time%s\t %s\n", _("[=seconds]"), _("generate a report for 'x' seconds"));
	printf(" -w, --workload%s %s\n", _("[=workload]"), _("file to execute for workload"));
	printf(" -V, --version\t\t %s\n", _("print version information"));
	printf(" -h, --help\t\t %s\n", _("print this help menu"));
	printf("\n");
	printf("%s\n\n", _("For more help please refer to the 'man 8 powertop'"));
}

static void do_sleep(int seconds)
{
	time_t target;
	int delta;

	if (!ncurses_initialized()) {
		sleep(seconds);
		return;
	}
	target = time(NULL) + seconds;
	delta = seconds;
	do {
		int c;
		usleep(6000);
		halfdelay(delta * 10);

		c = getch();
		switch (c) {
		case KEY_BTAB:
			show_prev_tab();
			break;
		case '\t':
			show_next_tab(); 
			break;
		case KEY_RIGHT:
			cursor_right(); 
			break;
		case KEY_LEFT:
			cursor_left(); 
			break;
		case KEY_NPAGE:
		case KEY_DOWN:
			cursor_down();
			break;
		case KEY_PPAGE:
		case KEY_UP:
			cursor_up();
			break;
		case ' ':
		case '\n':
			cursor_enter();
			break;
		case 's':
			if (set_refresh_timeout())
				return;
			break;
		case 'r':
			window_refresh();
			return;
		case KEY_EXIT:
		case 'q':
		case 27:	// Escape
			leave_powertop = 1;
			return;
		}

		delta = target - time(NULL);
		if (delta <= 0)
			break;

	} while (1);
}

extern "C" {
	static volatile bool end_thread;
	void* measure_background_thread(void *arg)
	{
		int sleep_time = *((int *) arg);
		while (!end_thread) {
			do_sleep(sleep_time);
			global_sample_power();
		}
		return 0;
	}
}

void one_measurement(int seconds, int sample_interval, char *workload)
{
	create_all_usb_devices();
	start_power_measurement();
	devices_start_measurement();
	start_devfreq_measurement();
	start_process_measurement();
	start_cpu_measurement();

	if (workload && workload[0]) {
		pthread_t thread = 0UL;
		end_thread = false;
		if (pthread_create(&thread, NULL, measure_background_thread, &sample_interval))
			fprintf(stderr, "ERROR: workload measurement thread creation failed\n");

		if (system(workload))
			fprintf(stderr, _("Unknown issue running workload!\n"));

		if (thread)
		{
			end_thread = true;
			pthread_join( thread, NULL);
		}
		global_sample_power();
	} else {
		while (seconds > 0)
		{
			do_sleep(sample_interval > seconds ? seconds : sample_interval);
			seconds -= sample_interval;
			global_sample_power();
		}
	}
	end_cpu_measurement();
	end_process_measurement();
	collect_open_devices();
	end_devfreq_measurement();
	devices_end_measurement();
	end_power_measurement();

	process_cpu_data();
	process_process_data();

	/* output stats */
	process_update_display();
	report_summary();
	w_display_cpu_cstates();
	w_display_cpu_pstates();
	if (reporttype != REPORT_OFF) {
		report_display_cpu_cstates();
		report_display_cpu_pstates();
	}
	report_process_update_display();
	tuning_update_display();
	wakeup_update_display();
	end_process_data();

	global_power();
	compute_bundle();

	show_report_devices();
	report_show_open_devices();

	report_devices();
	display_devfreq_devices();
	report_devfreq_devices();
	ahci_create_device_stats_table();
	store_results(measurement_time);
	end_cpu_data();
}

void out_of_memory()
{
	reset_display();
	printf("%s...\n",_("PowerTOP is out of memory. PowerTOP is Aborting"));
	abort();
}

void make_report(int time, char *workload, int iterations, int sample_interval, char *file)
{

	/* one to warm up everything */
	fprintf(stderr, _("Preparing to take measurements\n"));
	utf_ok = 0;
	one_measurement(1, sample_interval, NULL);

	if (!workload[0])
	  fprintf(stderr, _("Taking %d measurement(s) for a duration of %d second(s) each.\n"),iterations,time);
	else
	   fprintf(stderr, _("Measuring workload %s.\n"), workload);
	for (int i=0; i != iterations; i++){
		init_report_output(file, iterations);
		initialize_tuning();
		initialize_wakeup();
		/* and then the real measurement */
		one_measurement(time, sample_interval, workload);
		report_show_tunables();
		report_show_wakeup();
		finish_report_output();
		clear_tuning();
	}
	/* and wrap up */
	learn_parameters(50, 0);
	save_all_results("saved_results.powertop");
	save_parameters("saved_parameters.powertop");
	end_pci_access();
	exit(0);
}

static void checkroot() {
	int uid;
	uid = getuid();

	if (uid != 0) {
		printf(_("PowerTOP " PACKAGE_VERSION " must be run with root privileges.\n"));
		printf(_("exiting...\n"));
		exit(EXIT_FAILURE);
	}

}

static int get_nr_open(void) {
	int nr_open = NR_OPEN_DEF;
	ifstream file;

	file.open("/proc/sys/fs/nr_open", ios::in);
	if (file) {
		file >> nr_open;
		file.close();
	}
	return nr_open;
}

static void powertop_init(int auto_tune)
{
	static char initialized = 0;
	int ret;
	struct statfs st_fs;
	struct rlimit rlmt;

	if (initialized)
		return;

	checkroot();

	rlmt.rlim_cur = rlmt.rlim_max = get_nr_open();
	setrlimit (RLIMIT_NOFILE, &rlmt);

#if defined(__i386__) || defined(__x86_64__)
	if (system("/sbin/modprobe msr > /dev/null 2>&1"))
		fprintf(stderr, _("modprobe msr failed"));
#endif
	statfs("/sys/kernel/debug", &st_fs);

	if (st_fs.f_type != (long) DEBUGFS_MAGIC) {
		if (access("/bin/mount", X_OK) == 0) {
			ret = system("/bin/mount -t debugfs debugfs /sys/kernel/debug > /dev/null 2>&1");
		} else {
			ret = system("mount -t debugfs debugfs /sys/kernel/debug > /dev/null 2>&1");
		}
		if (ret != 0) {
			if (!auto_tune) {
				fprintf(stderr, _("Failed to mount debugfs!\n"));
				fprintf(stderr, _("exiting...\n"));
				exit(EXIT_FAILURE);
			} else {
				fprintf(stderr, _("Failed to mount debugfs!\n"));
				fprintf(stderr, _("Should still be able to auto tune...\n"));
			}
		}
	}

	srand(time(NULL));

	if (access("/var/cache/", W_OK) == 0)
		mkdir("/var/cache/powertop", 0600);
	else
		mkdir("/data/local/powertop", 0600);

	load_results("saved_results.powertop");
	load_parameters("saved_parameters.powertop");

	enumerate_cpus();
	create_all_devices();
	create_all_devfreq_devices();
	detect_power_meters();

	register_parameter("base power", 100, 0.5);
	register_parameter("cpu-wakeups", 39.5);
	register_parameter("cpu-consumption", 1.56);
	register_parameter("gpu-operations", 0.5576);
	register_parameter("disk-operations-hard", 0.2);
	register_parameter("disk-operations", 0.0);
	register_parameter("xwakes", 0.1);

        load_parameters("saved_parameters.powertop");

	initialized = 1;
}

void clean_shutdown()
{
	close_results();
	clean_open_devices();
	clear_all_devices();
	clear_all_devfreq();
	clear_all_cpus();

	return;
}


int main(int argc, char **argv)
{
	int option_index;
	int c;
	char filename[PATH_MAX];
	char workload[PATH_MAX] = {0};
	int  iterations = 1, auto_tune = 0, sample_interval = 5;

	set_new_handler(out_of_memory);

	setlocale (LC_ALL, "");

#ifdef ENABLE_NLS
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);
#endif
	ui_notify_user = ui_notify_user_ncurses;
	while (1) { /* parse commandline options */
		c = getopt_long(argc, argv, "cC::r::i:qt:w:Vh", long_options, &option_index);
		/* Detect the end of the options. */
		if (c == -1)
			break;
		switch (c) {
		case OPT_AUTO_TUNE:
			auto_tune = 1;
			leave_powertop = 1;
			ui_notify_user = ui_notify_user_console;
			break;
		case 'c':
			powertop_init(0);
			calibrate();
			break;
		case 'C':		/* csv report */
			reporttype = REPORT_CSV;
			snprintf(filename, sizeof(filename), "%s", optarg ? optarg : "powertop.csv");
			if (!strlen(filename))
			{
				fprintf(stderr, _("Invalid CSV filename\n"));
				exit(1);
			}
			break;
		case OPT_DEBUG:
			/* implemented using getopt_long(3) flag */
			break;
		case OPT_EXTECH:	/* Extech power analyzer support */
			checkroot();
			extech_power_meter(optarg ? optarg : "/dev/ttyUSB0");
			break;
		case 'r':		/* html report */
			reporttype = REPORT_HTML;
			snprintf(filename, sizeof(filename), "%s", optarg ? optarg : "powertop.html");
			if (!strlen(filename))
			{
				fprintf(stderr, _("Invalid HTML filename\n"));
				exit(1);
			}
			break;
		case 'i':
			iterations = (optarg ? atoi(optarg) : 1);
			break;
		case 'q':
			if (freopen("/dev/null", "a", stderr))
				fprintf(stderr, _("Quiet mode failed!\n"));
			break;
		case 's':
			sample_interval = (optarg ? atoi(optarg) : 5);
			break;
		case 't':
			time_out = (optarg ? atoi(optarg) : 20);
			break;
		case 'w':		/* measure workload */
			snprintf(workload, sizeof(workload), "%s", optarg ? optarg : "");
			break;
		case 'V':
			print_version();
			exit(0);
			break;
		case 'h':
			print_usage();
			exit(0);
			break;
		case '?':		/* Unknown option */
			/* getopt_long already printed an error message. */
			exit(1);
			break;
		}
	}

	powertop_init(auto_tune);

	if (reporttype != REPORT_OFF)
		make_report(time_out, workload, iterations, sample_interval, filename);

	if (debug_learning)
		printf("Learning debugging enabled\n");

	learn_parameters(250, 0);
	save_parameters("saved_parameters.powertop");


	if (debug_learning) {
	        learn_parameters(1000, 1);
		dump_parameter_bundle();
		end_pci_access();
		exit(0);
	}
	if (!auto_tune)
		init_display();

	initialize_devfreq();
	initialize_tuning();
	initialize_wakeup();
	/* first one is short to not let the user wait too long */
	one_measurement(1, sample_interval, NULL);

	if (!auto_tune) {
		tuning_update_display();
		show_tab(0);
	} else {
		auto_toggle_tuning();
	}

	while (!leave_powertop) {
		if (!auto_tune)
			show_cur_tab();
		one_measurement(time_out, sample_interval, NULL);
		learn_parameters(15, 0);
	}
	if (!auto_tune)
		endwin();
	fprintf(stderr, "%s\n", _("Leaving PowerTOP"));

	end_process_data();
	clear_process_data();
	end_cpu_data();
	clear_cpu_data();

	save_all_results("saved_results.powertop");
	save_parameters("saved_parameters.powertop");
	learn_parameters(500, 0);
	save_parameters("saved_parameters.powertop");
	end_pci_access();
	clear_tuning();
	reset_display();

	clean_shutdown();

	return 0;
}
